frame/staking-rewards/src/lib.rs

```diff
@@ -1051,58 +1118,121 @@ fn update_rewards_pool<T: Config>(
 	})
 }
 
-pub(crate) fn reward_accumulation_calculation<T: Config>(
-	reward: Reward<T::AssetId, T::Balance>,
+/// Calculates the update to the reward and unlocks the accumulated rewards from the pool account.
+pub(crate) fn do_reward_accumulation<T: Config>(
+	reward: &mut Reward<T::AssetId, T::Balance>,
+	pool_account: &T::AccountId,
 	now_seconds: u64,
-) -> Result<Reward<T::AssetId, T::Balance>, RewardAccumulationCalculationError<T>> {
-	match now_seconds.safe_sub(&reward.last_updated_timestamp) {
-		Ok(elapsed_time) => {
-			let reward_rate_period_seconds = reward.reward_rate.period.as_secs();
+) -> Result<(), RewardAccumulationCalculationError> {
+	if reward.reward_rate.amount.is_zero() {
+		return Ok(())
+	}
+
+	let elapsed_time = now_seconds
+		.safe_sub(&reward.last_updated_timestamp)
+		.map_err(|_| RewardAccumulationCalculationError::BackToTheFuture)?;
 
-			// SAFETY(benluelo): Usage of Div::div:
+	let reward_rate_period_seconds = reward.reward_rate.period.as_secs();
+
+	// SAFETY(benluelo): Usage of Div::div: Integer division can only fail if rhs == 0, and
+	// reward_rate_period_seconds is a NonZeroU64 here
+	let periods_surpassed = elapsed_time.div(reward_rate_period_seconds.get());
+
+	if periods_surpassed.is_zero() {
+		Ok(())
+	} else {
+		let total_locked_rewards: u128 =
+			T::Assets::balance_on_hold(reward.asset_id, pool_account).into();
+
+		// the maximum amount repayable given the reward rate.
+		// i.e. if total locked is 50, and the reward rate is 15, then this would be 3
+		//
+		// SAFETY(benluelo): Usage of Div::div: Integer division can only fail if rhs == 0, and
+		// reward.reward_rate.amount is known to be non-zero here as per the check at the beginning
+		// of this function
+		let maximum_releasable_periods = total_locked_rewards.div(reward.reward_rate.amount.into());
+
+		let releasable_periods_surpassed =
+			cmp::min(maximum_releasable_periods, periods_surpassed.into());
+
+		// saturating is safe here since these values are checked against max_rewards anyways, which
+		// is <= u128::MAX
+		let newly_accumulated_rewards =
+			u128::saturating_mul(releasable_periods_surpassed, reward.reward_rate.amount.into());
+		let new_total_rewards =
+			newly_accumulated_rewards.saturating_add(reward.total_rewards.into());
+
+		// u64::MAX is roughly 584.9 billion years in the future, so saturating at that should be ok
+		let last_updated_timestamp = reward.last_updated_timestamp.saturating_add(
+			reward_rate_period_seconds
+				.get()
+				.saturating_mul(releasable_periods_surpassed.saturated_into::<u64>()),
+		);
+
+		if new_total_rewards <= reward.max_rewards.into() {
```

Comment:
```suggestion
		if new_total_rewards <= (reward.max_rewards + reward.total_dilution_adjustment).into() {
```

Reply:
I added the dilution adjustment to the total rewards since max_rewards is documented as the total + dilution adjustment